Use connection URI in SqliteHook

feluelle opened the pull request:
We use the uri syntax for connecting to a sqlite database.
This allows the user to define more sqlite args such as mode. See https://docs.sqlalchemy.org/en/14/dialects/sqlite.html#uri-connections for details.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

dstandish commented on airflow/providers/sqlite/hooks/sqlite.py:
```diff
@@ -33,13 +34,16 @@ class SqliteHook(DbApiHook):
+        # See https://docs.sqlalchemy.org/en/14/dialects/sqlite.html#uri-connections for details.
+        uri = super().get_uri().replace("sqlite://", "sqlite:///")
+        raw_uri = unquote(uri)
+        return raw_uri
```
i assume this changes the output. will it break behavior for users?

dstandish replied:
i guess since it is a provider, we can break things, but you'll have to make it major release... i think the way to signal that to release manager is just add to changelog ?? @eladkal

eladkal replied:
If we can't deprecate first and must do a major release then we need to update provider.yaml major version to 4.0.0 and add relevant entry in the provider change log as part of this PR.

feluelle replied:
No, it does not change the output. We have tests which validates that [here](https://github.com/apache/airflow/pull/28721/files#diff-3614936e10a5d7aa9916f548b547cf3443f8ea29e30558d6fe55176e0a6f702eR30).

feluelle replied:
See line 139-140.

eladkal replied:
breaking change can also be about changing the input schema.
I did not review the PR, I just commented that if this is a breaking change then just need adding entry in change log and updating provider.yaml if this is not a breaking change but just a bug fix then no further action is required.

ashb reviewed (approved):
LGTM. Does it need docs https://airflow.apache.org/docs/apache-airflow-providers-sqlite/stable/connections/sqlite.html updating?

uranusjr commented on airflow/providers/sqlite/hooks/sqlite.py:
```diff
@@ -33,13 +34,20 @@ class SqliteHook(DbApiHook):
+        # The sqlite3 connection does not use the sqlite prefix.
+        # See https://docs.sqlalchemy.org/en/14/dialects/sqlite.html#connect-strings for details.
+        uri = self.get_uri().replace("sqlite:///", "file:")
+        conn = sqlite3.connect(uri, uri=True)
```
Is `:memory:` a viable value here? This logic does not support that.

feluelle replied:
Yes, it does. For example, see the last example [here](https://docs.python.org/3/library/sqlite3.html#how-to-work-with-sqlite-uris).
> db = "file:mem1?mode=memory&cache=shared"
> con1 = sqlite3.connect(db, uri=True)

feluelle replied:
More examples can be found [here](https://www.sqlite.org/inmemorydb.html).

uranusjr replied:
So you can’t use the `"sqlite://"` string (the default way to create an in-memory SQLite database in SQLAlchemy) since no way to configure a connection that formats into this specific string, but the same effect can be achieved using the URL format, right? I wonder if we want to document this somewhere.

feluelle replied:
You can. The host field is optional. You can create an empty SQLite connection and it will use an in-memory db.

feluelle replied:
I have added a test case for that 37d2a53.

And validated that it works locally:
```
>>> import sqlite3
>>> sqlite3.connect("file:")
<sqlite3.Connection object at 0xffffb71a4d50>
```

feluelle commented on airflow/providers/sqlite/hooks/sqlite.py:
```diff
@@ -33,13 +34,20 @@ class SqliteHook(DbApiHook):
+            airflow_conn.conn_type = self.conn_type
+        # The sqlite connection has one more slash for path specification.
+        # See https://docs.sqlalchemy.org/en/14/dialects/sqlite.html#uri-connections for details.
+        uri = airflow_conn.get_uri().replace("sqlite://", "sqlite:///")
```
FYI: This is also compatible with in-memory DB's in sqlalchemy.

For example:
```
>>> import sqlalchemy
>>> sqlalchemy.create_engine('sqlite:///').connect()
<sqlalchemy.engine.base.Connection object at 0xffffa95e5580>
```

ashb commented on docs/apache-airflow-providers-sqlite/connections/sqlite.rst:
```diff
@@ -23,14 +23,22 @@ The SQLite connection type provides connection to a SQLite database.
+
+.. code-block:: bash
+
+   export AIRFLOW_CONN_SQLITE_DEFAULT='sqlite://host/?mode=ro'
```
Host doesn't make any sense for a sqlite example

feluelle replied:
This is the path to the database. In this example the file "host" will be created in the current directory.

feluelle replied:
What do you want me to put there instead?

ashb replied:
Oh, have it as something like `sqlite:///path/to/db` then. And if it should only be `//` not `///` I think it's worth calling that out explicitly (as anyone familiar with sqlite might instinctively go for `///`)

ashb replied:
That said... Why are we changing it away from sqlite norms? Why can't we have `AIRFLOW_CONN_SQLITE_DEFAULT=sqlite:///path/to/db`? (Sorry, I've lost any and all context to this PR of late)

feluelle replied:
🤔 good question actually.. I will check tomorrow if we can remove the `replace` in `get_uri`.

feluelle replied:
I think it is because of `get_uri` impl in connection. It always has two slashes. So we need to add another one. Even if we add an absolute path in the host field we are still missing one. The absolute path in sql alchemy requires four slashes in total.

tldr; The URI you see in the docs is a connection URI - not a sqlalchemy uri.

ashb replied:
Thinking about it -- how does your PR behave if you try to specify a path to an absolute location? How would you create a URI for a db in `/tmp/my.sqlite`?

feluelle replied:
`sqlite:///tmp/my.sqlite`

feluelle replied:
You basically set the host field to the path and it will add another slash when getting the uri used for sql alchemy uris. For SQLite uris it will use `file:` instead.

feluelle replied:
Since this is quite confusing, I will add an example with a relative path and one with an absolute path.

ashb commented on docs/apache-airflow-providers-sqlite/connections/sqlite.rst:
```diff
@@ -23,14 +23,22 @@ The SQLite connection type provides connection to a SQLite database.
+
+   export AIRFLOW_CONN_SQLITE_DEFAULT='sqlite://host/?mode=ro'
+
+When specifying the connection as an environment variable in Airflow versions prior to 2.3.0, you need to specify the connection using the URI format.
```
What changed with 2.3? (I don't know what I was a user am meant to take from this paragraph)

feluelle replied:
This is a copy&paste from the snowflake one.

ashb replied:
Delete it then? It doesn't make any sense to me

Taragolis replied:
JSON format added: https://airflow.apache.org/docs/apache-airflow/stable/howto/connection.html#json-format-example

So better clarify or remove it

feluelle replied:
I have clarified it. :)

ashb commented on tests/providers/sqlite/hooks/test_sqlite.py:
```diff
@@ -28,26 +29,32 @@
-        self.db_hook.get_connection.return_value = self.connection
 
+    @pytest.mark.parametrize(
+        "connection, uri",
```
Nit:
```suggestion
        ["connection", "uri"],
```

feluelle replied:
This is not necessary.

Do you find your suggestion more readable?

Taragolis replied:
Yep, it not required for `pytest`, and in general we do not use it in our codebase

feluelle committed:
Use connection URI in SqliteHook

This allows the user to define more sqlite args such as mode. See https://docs.sqlalchemy.org/en/14/dialects/sqlite.html#uri-connections for details.
- remove unsupported schema, login and password fields in docs
- add info about host field to docs

feluelle commented on airflow/providers/sqlite/hooks/sqlite.py:
```diff
@@ -33,13 +34,21 @@ class SqliteHook(DbApiHook):
+        # The sqlite3 connection does not use the sqlite scheme and path should not end with a slash.
+        # See https://docs.sqlalchemy.org/en/14/dialects/sqlite.html#uri-connections for details.
+        sqlalchemy_uri = self.get_uri()
+        sqlite_uri = sqlalchemy_uri.replace("sqlite:///", "file:").replace("/?", "?")
```
The trailing slash in the path is a side effect of using `airflow_conn.get_uri()` 😞 I can also move the replace on this function call instead.

uranusjr replied:
It may be a good idea to move the logic to `get_uri`, this is arguably a bug.

feluelle replied:
I am not quite sure, because the URI format of an airflow connection is different from the one [SQLite uses](https://www.sqlite.org/uri.html).

feluelle replied:
SQLite `file:` URIs can't end with a slash because they are supposed to be files - not directories, but an airflow connection path can be anything.

feluelle committed:
Improve code readability
